src/Tools/Source/CompilerGeneratorTools/Source/CSharpSyntaxGenerator/Grammar/GrammarGenerator.cs

```diff
@@ -0,0 +1,325 @@
+                    // travelling too far away, and keeps the major sections relatively cohesive.
+                    var references = sorted.SelectMany(t => t.RuleReferences).Where(r => !s_majorRules.Contains(r));
+                    foreach (var reference in references)
+                        AddNormalizedRules(reference);
+                }
+            }
+        }
+
+        private static string Generate(string name, ImmutableArray<string> productions)
+            => Environment.NewLine + Environment.NewLine + name + Environment.NewLine + "  : " +
+               Join(Environment.NewLine + "  | ", productions) + Environment.NewLine + "  ;";
+
+        /// <summary>
+        /// Returns the g4 production string for this rule based on the children it has. Also
+        /// returns all the names of other rules this particular production references.
+        /// </summary>
+        private Production ProcessChildren(List<TreeTypeChild> children, string delim)
+        {
+            var result = children.Select(child => child switch
+            {
+                Choice c => ProcessChildren(c.Children, " | ").Parenthesize(),
+                Sequence s => ProcessChildren(s.Children, " ").Parenthesize(),
+                Field f => GetFieldType(f).WithSuffix(f.IsOptional ? "?" : ""),
+                _ => throw new InvalidOperationException(),
+            }).Where(p => p.Text.Length > 0);
+
+            return new Production(
+                Join(delim, result.Select(t => t.Text)),
+                result.SelectMany(t => t.RuleReferences));
+        }
+
+        private Production GetFieldType(Field field)
+            // 'bool' fields are for the few boolean properties we generate on DirectiveTrivia.
+            // They're not relevant to the grammar, so we just return an empty production here
+            // which will be filtered out by the caller.
+            => field.Type == "bool" ? new Production("") :
+               field.Type == "CSharpSyntaxNode" ? HandleCSharpSyntaxNodeField(field) :
+               field.Type.StartsWith("SeparatedSyntaxList") ? HandleSeparatedSyntaxListField(field) :
+               field.Type.StartsWith("SyntaxList") ? HandleSyntaxListField(field) :
+               field.IsToken ? HandleSyntaxTokenField(field) : RuleReference(field.Type);
+
+        private static Production HandleSyntaxTokenField(Field field)
+        {
+            var production = new Production(field.Kinds.Count == 0
+                ? GetTokenText(GetTokenKind(field.Name))
+                : Join(" | ", GetTokenKindStrings(field)));
+            return field.Kinds.Count <= 1 ? production : production.Parenthesize();
+        }
+
+        private Production HandleCSharpSyntaxNodeField(Field field)
+            => RuleReference(field.Kinds.Single().Name + Syntax);
+
+        private Production HandleSeparatedSyntaxListField(Field field)
```

Comment:
`static` (comment applies to a number of other methods)

Reply:
can't actually be static.  Ends up using RuleReference, which uses state from 'this'.

Reply:
Changed this up.  static now!